src/wh_client_crypto.c

```diff
@@ -1069,7 +1138,340 @@ int wh_Client_Curve25519SharedSecret(whClientContext* ctx,
 #endif
     return ret;
 }
-
 #endif /* HAVE_CURVE25519 */
 
+#ifndef NO_RSA
+int wh_Client_SetKeyIdRsa(RsaKey* key, whNvmId keyId)
+{
+    if (key == NULL)
+        return WH_ERROR_BADARGS;
+    key->devCtx = WH_KEYID_TO_DEVCTX(keyId);
+    return WH_ERROR_OK;
+}
+
+int wh_Client_GetKeyIdRsa(RsaKey* key, whNvmId* outId)
+{
+    if (key == NULL || outId == NULL)
+        return WH_ERROR_BADARGS;
+    *outId = WH_DEVCTX_TO_KEYID(key->devCtx);
+    return WH_ERROR_OK;
+}
+
+
+int wh_Client_ImportRsaKey(whClientContext* ctx, RsaKey* key,
+        whNvmFlags flags, uint32_t label_len, uint8_t* label,
+        whKeyId *out_keyId)
+{
+    int ret = 0;
+    whKeyId cacheKeyId = WH_KEYID_ERASED;
+    byte keyDer[WOLFHSM_CFG_COMM_DATA_LEN] = {0};
+    int derSize = 0;
+
+    if (    (ctx == NULL) ||
+            (key == NULL) ||
+            ((label_len != 0) && (label == NULL))) {
+        return WH_ERROR_BADARGS;
+    }
+
+    /* Convert RSA key to DER format */
+    ret = derSize = wc_RsaKeyToDer(key, keyDer, sizeof(keyDer));
+    if(derSize >= 0) {
+        /* Cache the key and get the keyID */
+        ret = wh_Client_KeyCache(ctx, flags,
+                label, label_len,
+                keyDer, derSize, &cacheKeyId);
+        if (out_keyId != NULL) {
+            *out_keyId = cacheKeyId;
+        }
+    }
+    return ret;
+}
+
+int wh_Client_ExportRsaKey(whClientContext* ctx, whKeyId keyId, RsaKey* key,
+        uint32_t label_len, uint8_t* label)
+{
+    int ret = 0;
+    /* DER cannot be larger than MTU */
+    byte keyDer[WOLFHSM_CFG_COMM_DATA_LEN] = {0};
+    uint16_t derSize = sizeof(keyDer);
+    uint8_t keyLabel[WH_NVM_LABEL_LEN] = {0};
+
+    if (    (ctx == NULL) ||
+            (keyId == WH_KEYID_ERASED) ||
+            (key == NULL)) {
+        return WH_ERROR_BADARGS;
+    }
+
+    /* Now export the key from the server */
+    ret = wh_Client_KeyExport(ctx,keyId,
+            keyLabel, sizeof(keyLabel),
+            keyDer, &derSize);
+    if (ret == 0) {
+        word32 idx = 0;
+        /* Update the RSA key structure */
+        ret = wc_RsaPrivateKeyDecode(
```

Comment:
same as above - this only handles private keys.

Reply:
Yup.  This should be updated to use deserialize.